src/transformers/models/superpoint/modeling_superpoint.py

```diff
@@ -0,0 +1,577 @@
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    logging,
+    replace_return_docstrings,
+)
+
+
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "SuperPointConfig"
+
+_CHECKPOINT_FOR_DOC = "stevenbucaille/superpoint"
+
+
+SUPERPOINT_PRETRAINED_MODEL_ARCHIVE_LIST = ["stevenbucaille/superpoint"]
+
+
+class SuperPointEncoder(nn.Module):
+    """
+    SuperPoint encoder module. It is made of 4 convolutional layers with ReLU activation and max pooling, reducing the
+     dimensionality of the image.
+    """
+
+    def __init__(self, config: SuperPointConfig) -> None:
+        super().__init__()
+        self.hidden_sizes = config.hidden_sizes
+        self.descriptor_dim = config.descriptor_dim
+
+        self.relu = nn.ReLU(inplace=True)
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+
+        self.conv1a = nn.Conv2d(1, self.hidden_sizes[0], kernel_size=3, stride=1, padding=1)
+        self.conv1b = nn.Conv2d(
+            self.hidden_sizes[0],
+            self.hidden_sizes[0],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv2a = nn.Conv2d(
+            self.hidden_sizes[0],
+            self.hidden_sizes[1],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv2b = nn.Conv2d(
+            self.hidden_sizes[1],
+            self.hidden_sizes[1],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv3a = nn.Conv2d(
+            self.hidden_sizes[1],
+            self.hidden_sizes[2],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv3b = nn.Conv2d(
+            self.hidden_sizes[2],
+            self.hidden_sizes[2],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv4a = nn.Conv2d(
+            self.hidden_sizes[2],
+            self.hidden_sizes[3],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv4b = nn.Conv2d(
+            self.hidden_sizes[3],
+            self.hidden_sizes[3],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+
+    def forward(
+        self,
+        input,
+        output_hidden_states: Optional[bool] = False,
+        return_dict: Optional[bool] = True,
+    ) -> Union[Tuple, BaseModelOutputWithNoAttention]:
+        all_hidden_states = () if output_hidden_states else None
+        input = self.relu(self.conv1a(input))
+        input = self.relu(self.conv1b(input))
+        input = self.pool(input)
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (input,)
+
+        input = self.relu(self.conv2a(input))
+        input = self.relu(self.conv2b(input))
+        input = self.pool(input)
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (input,)
```

Comment:
the repeated `if output_hidden_states` logic here indicates there should be a block abstraction which iterates over conv blocks rather than repeating the logic in the forward pass here

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/25786/commits/c43052de41132765b2bdfccb863a0ee7aa33a9c4)